examples/flower-via-docker-compose /helpers/generate_docker_compose.py

```diff
@@ -0,0 +1,127 @@
+import random
+
+
+def create_docker_compose(total_clients, number_of_rounds):
+    # cpus is used to set the number of CPUs available to the container as a fraction of the total number of CPUs on the host machine.
+    # mem_limit is used to set the memory limit for the container.
+    client_configs = [
+        {'mem_limit': '3g', 'batch_size': 32,  "cpus": 3.5, 'learning_rate': 0.001},
+        {'mem_limit': '4g', 'batch_size': 64,  "cpus": 3, 'learning_rate': 0.02},
+        {'mem_limit': '5g', 'batch_size': 128, "cpus": 2.5, 'learning_rate': 0.09},
+        {'mem_limit': '6g', 'batch_size': 256, "cpus": 1, 'learning_rate': 0.15},
+        # Add or modify the configurations depending on your host machine
+    ]
+
+    docker_compose_content = f"""
+version: '3'
+services:
+  prometheus:
+    image: prom/prometheus:latest
+    container_name: prometheus
+    ports:
+      - 9090:9090
+    mem_limit: 500m
+    command:
+      - --config.file=/etc/prometheus/prometheus.yml
+    volumes:
+      - ./config/prometheus.yml:/etc/prometheus/prometheus.yml:ro
+    depends_on:
+      - cadvisor
+    restart: always
+
+  cadvisor:
+    image: gcr.io/cadvisor/cadvisor:v0.47.0
+    container_name: cadvisor
+    restart: always
+    privileged: true
+    mem_limit: 500m
+    ports:
+      - "8080:8080"
+    volumes:
+      - /:/rootfs:ro
+      - /var/run:/var/run:ro
+      - /sys:/sys:ro
+      - /var/lib/docker/:/var/lib/docker:ro
+      - /dev/disk/:/dev/disk:ro
+      - /var/run/docker.sock:/var/run/docker.sock  
+
+  grafana:
+    image: grafana/grafana:latest
+    container_name: grafana
+    ports:
+      - 3000:3000
+    mem_limit: 400m
+    volumes:
+      - grafana-storage:/var/lib/grafana
+      - ./config/grafana.ini:/etc/grafana/grafana.ini
+      - ./config/provisioning/datasources:/etc/grafana/provisioning/datasources
+      - ./config/provisioning/dashboards:/etc/grafana/provisioning/dashboards
+    environment:
+      - GF_SECURITY_ADMIN_USER=admin
+      - GF_SECURITY_ADMIN_PASSWORD=admin
+    depends_on:
+      - prometheus
+      - cadvisor
+    command:
+      - --config=/etc/grafana/grafana.ini
+
+
+  server:
+    container_name: server
+    shm_size: '6g'
+    build:
+      context: .
+      dockerfile: Dockerfile
+    command: python server.py --number_of_rounds={number_of_rounds} --total_clients={total_clients}
+    restart: always
```

Comment:
The initial decision to use restart: always in the Docker Compose configuration was primarily to address situations where a client encountered an OOM error or similar issue, causing it to exit unexpectedly. This setting ensured that such clients would automatically restart, maintaining their participation in the training process. However, this approach led to a new challenge: after completing a training session (say, 10 rounds), the system would immediately initiate a new training cycle, which was not the intended behavior. To better align with our needs, it might be more appropriate to remove the restart: always policy altogether. This change would prevent automatic restarts of the training, allowing us more control over when and how training sessions begin.